### ISSUE TITLE
Update Lasagne PyPi package

### ISSUE THREAD

**Issue by cancan101** [Lasagne/Lasagne#575]
Any way to get more regular releases of Lasagne to pypi? It has been ~5 months since the last release even though there has been a lot of work on the repos.

**Comment by f0k** [comment]
> Any way to get more regular releases of Lasagne to pypi?

Well, we'd need to settle some discussions that might still change parts of the API of things added since the last release, as we don't want to change things in an incompatible way once they're released.
Off the top of my head, we need to resolve #399.
It seems as all the core developers use the bleeding-edge version, there's too little pressure to wrap things up and release -- probably the same problem as in Theano. Giving us some pressure will surely help :)

**Comment by baderbuddy** [comment]
Could the versions at least be updated as releases on GitHub even if it's not actually released to PyPi. If you keep the alpha tag then users will understand that the API can change.

**Comment by f0k** [comment]
> Could the versions at least be updated as releases on GitHub even if it's not actually released to PyPi.

We could -- what's the advantage / your use case?

**Comment by baderbuddy** [comment]
First of all, I'd like to say that I'm fully in favor of another alpha release to PyPi, even if the API changes that's fine, at least the latest version won't rely on a github version of Theano.

I think it'd be easier to sync up a team to make sure they're on the same version with releases and it gives people an option that want to use Lasagne without using the bleeding edge commit. With that said, the real reason is the organization I work at has a proxy that makes it really easy to get releases from GitHub and hard to get snapshots of an individual commit :)

**Comment by f0k** [comment]
> With that said, the real reason is the organization I work at has a proxy that makes it really easy to get releases from GitHub and hard to get snapshots of an individual commit :)

Ah, that's difficult to argue with :) For everything else, commit ids should work just as well (except they're far less user-friendly).

> at least the latest version won't rely on a github version of Theano.

This might quickly change again. But yes, maybe we should take the opportunity to make a Lasagne release that works with plain Theano 0.8.

**Comment by benanne** [comment]
Let's do it! What needs to happen for that? Are there any glaring bugs/issues we need to sort out first? We can do a little sprint :)

**Comment by f0k** [comment]
> Are there any glaring bugs/issues we need to sort out first?

There are some things we might want to change the interface for before releasing. As mentioned above, we were discussing #399. Batch normalization could be a candidate as well -- if we integrate cuDNN v5's interface in Theano, we may want to change the default interface for the BatchNormLayer (i.e., have it remember the variance instead of the inverse standard deviation). And we should probably add `output_size` to the transposed conv layer. The new layer base class and recurrent layers should be postponed to 0.3.
Other than that, we should review open issues and PRs and tag them with a 0.2 milestone if needed.

**Comment by benanne** [comment]
Sounds good to me.

**Comment by f0k** [comment]
> Other than that, we should review open issues and PRs and tag them with a 0.2 milestone if needed.

I've tagged everything that would be good to have or solve for 0.2 (not everything of this is a necessity, but I didn't distinguish this). I've also created a 0.3 milestone for things we shouldn't tackle before 0.2.

> Batch normalization could be a candidate as well -- if we integrate cuDNN v5's interface in Theano, we may want to change the default interface for the BatchNormLayer (i.e., have it remember the variance instead of the inverse standard deviation).

The cuDNN v5 interface is a bit weird in this respect -- it still returns the inverse standard deviation during the forward pass to be reused in the backward pass, but it also allows to update the non-inverted standard deviation, and the inference implementation requires the non-inverted standard deviation. A possible solution would be to use cuDNN just to obtain the batch-normalized data, its mean and its inverse standard deviation in one go (and also use cuDNN for the backward pass), but handle inference and the exponential moving average on our own. I've implemented the necessary Ops for Theano (Theano/Theano/pull/4582). We can only integrate this in Lasagne past v0.2 if we still want v0.2 to be compatible with Theano 0.8, but it means we can keep the BatchNormLayer interface as it is.

**Comment by benanne** [comment]
Excellent, the list of things to do for v0.2 looks fairly manageable, but the overall level of developer activity (myself included) has tapered off a bit. I'll try to set aside some time to tackle these in the next month or so.
